torch/_inductor/fx_passes/comm_fusion.py

```diff
@@ -0,0 +1,154 @@
+import logging
+import operator
+from typing import Any, Callable, Optional, Union
+
+import torch
+from torch._dynamo.utils import counters
+from torch._utils_internal import print_graph
+
+from ..pattern_matcher import (
+    CallFunctionVarArgs,
+    get_arg_value,
+    MULTIPLE,
+    PatternMatcherPass,
+    register_graph_pattern,
+)
+
+logger = logging.getLogger(__name__)
+
+
+comm_fusion_patterns = PatternMatcherPass()
+
+
+def comm_fusion_passes(graph: torch.fx.Graph) -> None:
+    """Pattern matching passes for collective communication fucntion fusion."""
+    if not torch.distributed.is_available:
+        logger.info(
+            "Warning: torch.distributed is not available, skip communication fusion"
+        )
+        return
+
+    print_graph(graph, "Before communication fusion in post grad pass.")
+    comm_fusion_patterns.apply(graph)
+    print_graph(graph, "After communication fusion in post grad pass.")
+
+
+@register_graph_pattern(
+    CallFunctionVarArgs(torch.ops.c10d_functional.all_reduce.default, users=MULTIPLE),
+    pass_dict=comm_fusion_patterns,
+)
+def allreduce_split_to_reducescatter(match, *args, **kwargs):
+    """
+    Replace the following pattern:
+        AllReduce() - wait() - split - getitem
+
+    with:
+        ReduceScatter() - wait()
+
+    """
+    node = match.nodes[0]
+    graph = match.graph
+    if len(node.args) != 5:
+        logger.warning("AllReduce node has %s args, expected 5", len(node.args))
+        return
+
+    # Match the pattern, for now the it only works with single user/child node.
+    wait_node = CommFusionHelper.get_only_child(
+        node, "call_function", torch.ops.c10d_functional.wait_tensor.default
+    )
+    if wait_node is None:
+        return
+
+    split_node = CommFusionHelper.get_only_child(
+        wait_node, "call_function", torch.ops.aten.split.Tensor
+    )
+    if split_node is None:
+        return
+
+    getitem_node = CommFusionHelper.get_only_child(
```

Comment:
Can you add some comment here to mention that checking only one getitem is required because DDP fusion has exactly the same pattern except DDP fusion has more than one geittem?